### ISSUE TITLE
Consistent segfaults since hstr 2.2.0

### ISSUE THREAD

**Issue by thibaultcha** [dvorka/hstr#408]
Hi there!

For the last 20 minutes or so, I am running into a segfault each time I press Ctrl-C when hstr is open:

```
$ hh
free(): invalid pointer
[1]    2797722 abort (core dumped)  hh
```

```
$ hstr -V
hstr version "2.2.0" (2019-12-21T08:47:00)
$ uname -a
Linux x1c 5.6.16-300.fc32.x86_64 #1 SMP Thu Jun 4 18:08:38 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

Here is a gdb session:
```
           PID: 2796467 (hh)
           UID: 1000 (...)
           GID: 1000 (...)
        Signal: 6 (ABRT)
     Timestamp: Wed 2020-06-17 19:06:25 PDT (6s ago)
  Command Line: hh
    Executable: /usr/bin/hh
 Control Group: /user.slice/user-1000.slice/session-2.scope
          Unit: session-2.scope
         Slice: user-1000.slice
       Session: 2
     Owner UID: 1000 (...)
       Boot ID: 9ca955e526fc477a89743e1703948c21
    Machine ID: 18bea806d06c49a29c1b5cd159b7ed73
      Hostname: ...
       Storage: /var/lib/systemd/coredump/core.hh.1000.9ca955e526fc477a89743e1703948c21.2796467.1592445985000000000000.lz4
       Message: Process 2796467 (hh) of user 1000 dumped core.
                Stack trace of thread 2796467:
                #0  0x00007f3f69e4da25 __GI_raise (libc.so.6 + 0x3ca25)
                #1  0x00007f3f69e36895 __GI_abort (libc.so.6 + 0x25895)
                #2  0x00007f3f69e918c7 __libc_message (libc.so.6 + 0x808c7)
                #3  0x00007f3f69e98dec malloc_printerr (libc.so.6 + 0x87dec)
                #4  0x00007f3f69e99fb4 _int_free (libc.so.6 + 0x88fb4)
                #5  0x0000560cf8a4d7f6 prioritized_history_destroy (hh + 0x37f6)
                #6  0x0000560cf8a4fad6 hstr_destroy (hh + 0x5ad6)
                #7  0x0000560cf8a4fb1c hstr_exit (hh + 0x5b1c)
                #8  0x0000560cf8a4fb51 signal_callback_handler_ctrl_c (hh + 0x5b51)
                #9  0x00007f3f69e4dab0 __restore_rt (libc.so.6 + 0x3cab0)
                #10 0x00007f3f69f03540 __GI___libc_read (libc.so.6 + 0xf2540)
                #11 0x00007f3f6a01f369 _nc_wgetch (libncursesw.so.6 + 0x15369)
                #12 0x00007f3f6a01ff0b wgetch (libncursesw.so.6 + 0x15f0b)
                #13 0x0000560cf8a51e64 loop_to_select (hh + 0x7e64)
                #14 0x0000560cf8a52ce6 hstr_interactive (hh + 0x8ce6)
                #15 0x0000560cf8a52efd hstr_main (hh + 0x8efd)
                #16 0x00007f3f69e38042 __libc_start_main (libc.so.6 + 0x27042)
                #17 0x0000560cf8a4ca4e _start (hh + 0x2a4e)

GNU gdb (GDB) Fedora 9.1-5.fc32
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/bin/hh...
Reading symbols from /usr/lib/debug/usr/bin/hh-2.2-2.fc32.x86_64.debug...
[New LWP 2796467]

Program terminated with signal SIGABRT, Aborted.

#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50        return ret;
>>> frame 5
#5  0x000055d136b767f6 in prioritized_history_destroy (h=0x55d138b547a0) at hstr_history.c:259
259                         free(h->items[i]);
>>> print *h
$1 = {
  items = 0x55d138b75430,
  count = 889,
  rawItems = 0x55d138af42c0,
  rawCount = 2112
}
>>> print i
$2 = 329
```

I hope that helps.

**Comment by thibaultcha** [comment]
An update on this: I am also noticing the same segfaults when I press Tab to select a past command, and not only on Ctrl-C.

**Comment by thibaultcha** [comment]
Hi @dvorka,

Have you had a chance to look at this issue? Over the past month, these segfaults have been relentlessly consistent and keep happening at every single invocation of hstr. It seems like other people are experiencing them as well, see #399.

**Comment by thibaultcha** [comment]
Turned our the issue was because of a command I mispelled which started with a `:`, which would confuse the zsh extended history parsing. This ended-up being quite fun to track down.

Fixed in #410. Closing this issue since it the root cause has nothing to do with hstr 2.2.0.
